File: app_manager/appmanager/src/main/java/app/editors/manager/managers/providers/CloudFileProvider.java
package app.editors.manager.managers.providers;

import android.net.Uri;

import androidx.annotation.Nullable;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.Map;

import app.editors.manager.app.Api;
import app.editors.manager.mvp.models.base.Base;
import app.editors.manager.mvp.models.explorer.Explorer;
import app.editors.manager.mvp.models.explorer.File;
import app.editors.manager.mvp.models.explorer.Folder;
import app.editors.manager.mvp.models.explorer.Item;
import app.editors.manager.mvp.models.explorer.Operation;
import app.editors.manager.mvp.models.request.RequestBatchBase;
import app.editors.manager.mvp.models.request.RequestBatchOperation;
import app.editors.manager.mvp.models.request.RequestCreate;
import app.editors.manager.mvp.models.request.RequestExternal;
import app.editors.manager.mvp.models.request.RequestFavorites;
import app.editors.manager.mvp.models.request.RequestRenameFile;
import app.editors.manager.mvp.models.request.RequestTitle;
import app.editors.manager.mvp.models.response.ResponseExternal;
import app.editors.manager.mvp.models.response.ResponseOperation;
import io.reactivex.Observable;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.schedulers.Schedulers;
import retrofit2.HttpException;

public class CloudFileProvider implements BaseFileProvider {

    public enum Section {
        My("@my"),
        Common("@common"),
        Shared("@share"),
        Projects("@projects"),
        Trash("@trash"),
        Favorites("@favorites");

        String mPath;

        Section(String path) {
            this.mPath = path;
        }

        public String getPath() {
            return mPath;
        }
    }

    private String mToken;
    private Api mApi;

    public CloudFileProvider(String mToken, Api mApi) {
        this.mToken = mToken;
        this.mApi = mApi;
    }

    @Override
    public Observable<Explorer> getFiles(String id, @Nullable Map<String, String> filter) {
        return Observable.fromCallable(() -> mApi.getItemById(mToken, id, filter).execute())
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .map(responseExplorerResponse -> {
                    if (responseExplorerResponse.isSuccessful() && responseExplorerResponse.body() != null) {
                        return responseExplorerResponse.body().getResponse();
                    } else {
                        throw new HttpException(responseExplorerResponse);
                    }
                });
    }

    @Override
    public Observable<File> createFile(String folderId, RequestCreate body) {
        return Observable.fromCallable(() -> mApi.createDocs(mToken, folderId, body).execute())
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .map(responseCreateFile -> {
                    if (responseCreateFile.isSuccessful() && responseCreateFile.body() != null) {
                        return responseCreateFile.body().getResponse();
                    } else {
                        throw new HttpException(responseCreateFile);
                    }
                });
    }

    @Override
    public Observable<Folder> createFolder(String folderId, RequestCreate body) {
        return Observable.fromCallable(() -> mApi.createFolder(mToken, folderId, body).execute())
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .map(responseCreateFolder -> {
                    if (responseCreateFolder.isSuccessful() && responseCreateFolder.body() != null) {
                        return responseCreateFolder.body().getResponse();
                    } else {
                        throw new HttpException(responseCreateFolder);
                    }
                });
    }

    @Override
    public Observable<Item> rename(Item item, String newName, @Nullable Integer version) {
        if (version == null) {
            return folderRename(item.getId(), newName);
        } else {
            return fileRename(item.getId(), newName, version);
        }
    }

    private Observable<Item> fileRename(String id, String newName, Integer version) {
        RequestRenameFile requestRenameFile = new RequestRenameFile();
        requestRenameFile.setLastVersion(version);
        requestRenameFile.setTitle(newName);
        return Observable.fromCallable(() -> mApi.renameFile(mToken, id, requestRenameFile).execute())
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .map(responseFile -> {
                    if (responseFile.isSuccessful() && responseFile.body() != null) {
                        return responseFile.body().getResponse();
                    } else if (responseFile.code() == 400) {
                        throw ProviderError.throwForbiddenError();
                    } else {
                        throw new HttpException(responseFile);
                    }
                });
    }

    private Observable<Item> folderRename(String id, String newName) {
        RequestTitle requestTitle = new RequestTitle();
        requestTitle.setTitle(newName);
        return Observable.fromCallable(() -> mApi.renameFolder(mToken, id, requestTitle).execute())
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .map(responseFolder -> {
                    if (responseFolder.isSuccessful() && responseFolder.body() != null) {
                        return responseFolder.body().getResponse();
                    } else if (responseFolder.code() == 400) {
                        throw ProviderError.throwForbiddenError();
                    } else {
                        throw new HttpException(responseFolder);
                    }
                });
    }

    @Override
    public Observable<List<Operation>> delete(List<Item> items, @Nullable Folder from) {
        List<String> filesId = new ArrayList<>();
        List<String> foldersId = new ArrayList<>();
        for (Item item : items) {
            if (item instanceof File) {
                filesId.add(item.getId());
            } else if (item instanceof Folder) {
                if (item.getProviderItem()) {
                    removeStorage(item.getId());
                } else {
                    foldersId.add(item.getId());
                }
            }
        }
        RequestBatchBase request = new RequestBatchBase();
        request.setDeleteAfter(false);
        request.setFileIds(filesId);
        request.setFolderIds(foldersId);
        return Observable.fromCallable(() -> mApi.deleteBatch(mToken, request).execute())
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .map(responseOperationResponse -> {
                    if (responseOperationResponse.isSuccessful() && responseOperationResponse.body() != null) {
                        return responseOperationResponse.body().getResponse();
                    } else {
                        throw new HttpException(responseOperationResponse);
                    }
                });
    }

    private void removeStorage(String id) {
        Observable.fromCallable(() -> mApi.deleteStorage(mToken, id.substring(id.indexOf('-') + 1)).execute())
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .map(responseBody -> {
                    if (responseBody.isSuccessful()) {
                        Operation operation = new Operation();
                        operation.setProgress(100);
                        return Collections.singletonList(operation);
                    } else {
                        throw new HttpException(responseBody);
                    }
                }).subscribe();
    }

    @Override
    public Observable<Integer> download(List<Item> items) {
        return null;
    }

    @Override
    public Observable<Integer> upload(String folderId, List<Uri> uri) {
        return null;
    }

    @Override
    public Observable<List<Operation>> transfer(List<Item> items, @Nullable Folder to, int conflict, boolean isMove, boolean isOverwrite) {
        List<String> filesId = new ArrayList<>();
        List<String> foldersId = new ArrayList<>();
        for (Item item : items) {
            if (item instanceof File) {
                filesId.add(item.getId());
            } else if (item instanceof Folder) {
                foldersId.add(item.getId());
            }
        }

        RequestBatchOperation batchOperation = new RequestBatchOperation();
        batchOperation.setFileIds(filesId);
        batchOperation.setFolderIds(foldersId);
        batchOperation.setDeleteAfter(false);
        batchOperation.setDestFolderId(to.getId());
        batchOperation.setConflictResolveType(conflict);

        if (isMove) {
            return moveItems(batchOperation);
        } else {
            return copyFiles(batchOperation);
        }
    }

    private Observable<List<Operation>> moveItems(RequestBatchOperation body) {
        return Observable.fromCallable(() -> mApi.move(mToken, body).execute())
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .map(responseOperation -> {
                    if (responseOperation.isSuccessful() && responseOperation.body() != null) {
                        return responseOperation.body().getResponse();
                    } else {
                        throw new HttpException(responseOperation);
                    }
                });
    }

    private Observable<List<Operation>> copyFiles(RequestBatchOperation body) {
        return Observable.fromCallable(() -> mApi.copy(mToken, body).execute())
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .map(responseOperation -> {
                    if (responseOperation.isSuccessful() && responseOperation.body() != null) {
                        return responseOperation.body().getResponse();
                    } else {
                        throw new HttpException(responseOperation);
                    }
                });
    }

    @Override
    public Observable<File> fileInfo(Item item) {
        return Observable.fromCallable(() -> mApi.getFileInfo(mToken, item.getId()).execute())
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .map(responseFile -> {
                    if (responseFile.isSuccessful() && responseFile.body() != null) {
                        return responseFile.body().getResponse();
                    } else {
                        throw new HttpException(responseFile);
                    }
                });
    }

    @Override
    public ResponseOperation getStatusOperation() {
        return mApi.status(mToken)
                .blockingGet();
    }

    @Override
    public Observable<ResponseExternal> share(String id, RequestExternal requestExternal) {
        return Observable.fromCallable(() -> mApi.getExternalLink(mToken, id, requestExternal).execute())
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .map(responseExternal -> {
                    if (responseExternal.isSuccessful()) {
                        return responseExternal.body();
                    } else {
                        throw new HttpException(responseExternal);
                    }
                });
    }

    @Override
    public Observable<List<Operation>> terminate() {
        return Observable.fromCallable(() -> mApi.terminate(mToken).execute())
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .map(operationResponse -> {
                    if (operationResponse.isSuccessful() && operationResponse.body() != null) {
                        return operationResponse.body().getResponse();
                    } else {
                        throw new HttpException(operationResponse);
                    }
                });
    }

    @Override
    public Observable<Base> addToFavorites(RequestFavorites requestFavorites) {
        return Observable.fromCallable(() -> mApi.addToFavorites(mToken, requestFavorites).execute())
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .map(baseResponse -> {
                    return baseResponse.body();
                });
    }

    @Override
    public Observable<Base> deleteFromFavorites(RequestFavorites requestFavorites) {
        return Observable.fromCallable(() -> mApi.deleteFromFavorites(mToken, requestFavorites).execute())
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .map(baseResponse -> {
                    return baseResponse.body();
                });
    }

    public Observable<List<Operation>> clearTrash() {
        return Observable.fromCallable(() -> mApi.emptyTrash(mToken).execute())
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .map(operationResponse -> {
                    if (operationResponse.isSuccessful() && operationResponse.body() != null) {
                        return operationResponse.body().getResponse();
                    } else {
                        throw new HttpException(operationResponse);
                    }
                });
    }

}

File: app_manager/appmanager/src/main/java/app/editors/manager/managers/providers/CloudFileProvider.kt
package app.editors.manager.managers.providers

import android.net.Uri
import app.editors.manager.app.Api
import app.editors.manager.app.App
import app.editors.manager.mvp.models.base.Base
import app.editors.manager.mvp.models.explorer.*
import app.editors.manager.mvp.models.request.*
import app.editors.manager.mvp.models.response.*
import io.reactivex.Observable
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.schedulers.Schedulers
import okhttp3.ResponseBody
import retrofit2.HttpException
import retrofit2.Response
import java.util.*

class CloudFileProvider : BaseFileProvider {

    var api: Api = App.getApp().getApi()

    override fun getFiles(id: String?, filter: Map<String, String>?): Observable<Explorer> {
        return id?.let {
            api.getItemById(it, mapOf())
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .map { responseExplorerResponse: Response<ResponseExplorer> ->
                    if (responseExplorerResponse.isSuccessful && responseExplorerResponse.body() != null) {
                        return@map responseExplorerResponse.body()!!.response
                    } else {
                        throw HttpException(responseExplorerResponse)
                    }
                }
        }!!
    }

    override fun search(query: String?): Observable<String>? {
        return query?.let {
            api.search(it)
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .map { responseSearchResponse ->
                    if (responseSearchResponse.isSuccessful && responseSearchResponse.body() != null) {
                        return@map responseSearchResponse.body()!!.string()
                    } else {
                        throw HttpException(responseSearchResponse)
                    }
                }
        }
    }

    override fun createFile(folderId: String, body: RequestCreate): Observable<CloudFile> {
        return api.createDocs(folderId, body)
            .subscribeOn(Schedulers.io())
            .observeOn(AndroidSchedulers.mainThread())
            .map { responseCreateFile: Response<ResponseCreateFile> ->
                if (responseCreateFile.isSuccessful && responseCreateFile.body() != null) {
                    return@map responseCreateFile.body()!!.response
                } else {
                    throw HttpException(responseCreateFile)
                }
            }
    }

    override fun createFolder(folderId: String, body: RequestCreate): Observable<CloudFolder> {
        return api.createFolder(folderId, body)
            .subscribeOn(Schedulers.io())
            .observeOn(AndroidSchedulers.mainThread())
            .map { responseCreateFolder: Response<ResponseCreateFolder> ->
                if (responseCreateFolder.isSuccessful && responseCreateFolder.body() != null) {
                    return@map responseCreateFolder.body()!!.response
                } else {
                    throw HttpException(responseCreateFolder)
                }
            }
    }

    override fun rename(item: Item, newName: String, version: Int?): Observable<Item> {
        return if (version == null) {
            folderRename(item.id, newName)
        } else {
            fileRename(item.id, newName, version)
        }
    }

    private fun fileRename(id: String, newName: String, version: Int): Observable<Item> {
        val requestRenameFile = RequestRenameFile()
        requestRenameFile.lastVersion = version
        requestRenameFile.title = newName
        return api.renameFile(id, requestRenameFile)
            .subscribeOn(Schedulers.io())
            .observeOn(AndroidSchedulers.mainThread())
            .map { responseFile: Response<ResponseFile> ->
                if (responseFile.isSuccessful && responseFile.body() != null) {
                    return@map responseFile.body()!!.response
                } else if (responseFile.code() == 400) {
                    throw ProviderError.throwForbiddenError()
                } else {
                    throw HttpException(responseFile)
                }
            }
    }

    private fun folderRename(id: String, newName: String): Observable<Item> {
        val requestTitle = RequestTitle()
        requestTitle.title = newName
        return api.renameFolder(id, requestTitle)
            .subscribeOn(Schedulers.io())
            .observeOn(AndroidSchedulers.mainThread())
            .map { responseFolder: Response<ResponseFolder> ->
                if (responseFolder.isSuccessful && responseFolder.body() != null) {
                    return@map responseFolder.body()!!.response
                } else if (responseFolder.code() == 400) {
                    throw ProviderError.throwForbiddenError()
                } else {
                    throw HttpException(responseFolder)
                }
            }
    }

    override fun delete(items: List<Item>, from: CloudFolder?): Observable<List<Operation>> {
        val filesId: MutableList<String> = ArrayList()
        val foldersId: MutableList<String> = ArrayList()
        for (item in items) {
            if (item is CloudFile) {
                filesId.add(item.getId())
            } else if (item is CloudFolder) {
                if (item.getProviderItem()) {
                    removeStorage(item.getId())
                } else {
                    foldersId.add(item.getId())
                }
            }
        }
        val request = RequestBatchBase()
        request.isDeleteAfter = false
        request.fileIds = filesId
        request.folderIds = foldersId
        return Observable.fromCallable {
            api.deleteBatch(
                request
            ).execute()
        }
            .subscribeOn(Schedulers.io())
            .observeOn(AndroidSchedulers.mainThread())
            .map { responseOperationResponse: Response<ResponseOperation?> ->
                if (responseOperationResponse.isSuccessful && responseOperationResponse.body() != null) {
                    return@map responseOperationResponse.body()!!.response
                } else {
                    throw HttpException(responseOperationResponse)
                }
            }
    }

    private fun removeStorage(id: String) {
        api.deleteStorage(id.substring(id.indexOf('-') + 1))
            .subscribeOn(Schedulers.io())
            .observeOn(AndroidSchedulers.mainThread())
            .map { responseBody: Response<ResponseBody> ->
                if (responseBody.isSuccessful) {
                    val operation = Operation()
                    operation.progress = 100
                    return@map listOf(operation)
                } else {
                    throw HttpException(responseBody)
                }
            }.subscribe()
    }

    override fun download(items: List<Item>): Observable<Int>? {
        return null
    }

    override fun upload(folderId: String, uris: List<Uri?>): Observable<Int>? {
        return null
    }

    override fun transfer(
        items: List<Item>,
        to: CloudFolder?,
        conflict: Int,
        isMove: Boolean,
        isOverwrite: Boolean
    ): Observable<List<Operation>> {
        val filesId: MutableList<String> = ArrayList()
        val foldersId: MutableList<String> = ArrayList()
        for (item in items) {
            if (item is CloudFile) {
                filesId.add(item.getId())
            } else if (item is CloudFolder) {
                foldersId.add(item.getId())
            }
        }
        val batchOperation = RequestBatchOperation()
        batchOperation.fileIds = filesId
        batchOperation.folderIds = foldersId
        batchOperation.isDeleteAfter = false
        batchOperation.destFolderId = to!!.id
        batchOperation.conflictResolveType = conflict
        return if (isMove) {
            moveItems(batchOperation)
        } else {
            copyFiles(batchOperation)
        }
    }

    private fun moveItems(body: RequestBatchOperation): Observable<List<Operation>> {
        return api.move(body)
            .subscribeOn(Schedulers.io())
            .observeOn(AndroidSchedulers.mainThread())
            .map { responseOperation: Response<ResponseOperation> ->
                if (responseOperation.isSuccessful && responseOperation.body() != null) {
                    return@map responseOperation.body()!!.response
                } else {
                    throw HttpException(responseOperation)
                }
            }
    }

    private fun copyFiles(body: RequestBatchOperation): Observable<List<Operation>> {
        return api.copy(body)
            .subscribeOn(Schedulers.io())
            .observeOn(AndroidSchedulers.mainThread())
            .map { responseOperation: Response<ResponseOperation> ->
                if (responseOperation.isSuccessful && responseOperation.body() != null) {
                    return@map responseOperation.body()!!.response
                } else {
                    throw HttpException(responseOperation)
                }
            }
    }

    override fun fileInfo(item: Item?): Observable<CloudFile> {
        return api.getFileInfo(item?.id)
            .subscribeOn(Schedulers.io())
            .observeOn(AndroidSchedulers.mainThread())
            .map { responseFile: Response<ResponseFile> ->
                if (responseFile.isSuccessful && responseFile.body() != null) {
                    return@map responseFile.body()!!.response
                } else {
                    throw HttpException(responseFile)
                }
            }
    }

    override fun getStatusOperation(): ResponseOperation {
        return api.status()
            .blockingGet()
    }

    override fun share(id: String, requestExternal: RequestExternal): Observable<ResponseExternal> {
        return api.getExternalLink(id, requestExternal)
            .subscribeOn(Schedulers.io())
            .observeOn(AndroidSchedulers.mainThread())
            .map { responseExternal: Response<ResponseExternal> ->
                if (responseExternal.isSuccessful) {
                    return@map responseExternal.body()
                } else {
                    throw HttpException(responseExternal)
                }
            }
    }

    override fun terminate(): Observable<List<Operation>> {
        return Observable.fromCallable {
            api.terminate().execute()
        }
            .subscribeOn(Schedulers.io())
            .observeOn(AndroidSchedulers.mainThread())
            .map { operationResponse: Response<ResponseOperation?> ->
                if (operationResponse.isSuccessful && operationResponse.body() != null) {
                    return@map operationResponse.body()!!.response
                } else {
                    throw HttpException(operationResponse)
                }
            }
    }

    override fun addToFavorites(requestFavorites: RequestFavorites): Observable<Base> {
        return api.addToFavorites(requestFavorites)
            .subscribeOn(Schedulers.io())
            .observeOn(AndroidSchedulers.mainThread())
            .map { baseResponse: Response<Base> -> baseResponse.body() }
    }

    override fun deleteFromFavorites(requestFavorites: RequestFavorites): Observable<Base> {
        return api.deleteFromFavorites(requestFavorites)
            .subscribeOn(Schedulers.io())
            .observeOn(AndroidSchedulers.mainThread())
            .map { baseResponse: Response<Base> -> baseResponse.body() }
    }

    fun clearTrash(): Observable<List<Operation>> {
        return api.emptyTrash()
            .subscribeOn(Schedulers.io())
            .observeOn(AndroidSchedulers.mainThread())
            .map { operationResponse: Response<ResponseOperation> ->
                if (operationResponse.isSuccessful && operationResponse.body() != null) {
                    return@map operationResponse.body()!!.response
                } else {
                    throw HttpException(operationResponse)
                }
            }
    }
}